### PR TITLE
plugins/frustum_culling: Add frustum culling plugin

### PR DESCRIPTION
**Pull request by mats4**

### CODE REVIEW

File: plugins/frustum_culling/src/frustum_culling_conf.hpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+#include <cloe/component.hpp>          // for Component, ComponentFactory, ...
+#include <cloe/component/frustum.hpp>  // for Frustum
+#include <cloe/core.hpp>               // for Confable, Schema
+#include <cloe/entity.hpp>             // for Entity
+#include <cloe/simulator.hpp>          // for ModelError
+#include <cloe/utility/geometry.hpp>   // for quaternion_from_rpy
+#include <fable/json/with_eigen.hpp>   // for to_json
+
+namespace cloe {
+namespace component {
```
[the remark this answers is not in the document]
Comment (cassava) [reply plugins/frustum_culling/src/frustum_culling_conf.hpp:38]: Or better yet, `frustum_culling_plugin` combined with a `using namespace cloe;`.

**Review by cassava — CHANGES_REQUESTED** [review]

File: models/include/cloe/utility/frustum_culling.hpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  double x{0.0};
+  double y{0.0};
+};
+
+// Rotates a given point from one coordinate system to another and returns rotated point
+Point rotate_point(const Point& point, double angle) {
+  Point p{0.0, 0.0};
+  p.x = std::cos(angle) * point.x - std::sin(angle) * point.y;
+  p.y = std::sin(angle) * point.x + std::cos(angle) * point.y;
+  return p;
```
Comment (cassava) [inline models/include/cloe/utility/frustum_culling.hpp:44]:
```suggestion
  return Point{
    std::cos(angle) * point.x - std::sin(angle) * point.y,
    std::sin(angle) * point.x + std::cos(angle) * point.y,
  };
```

File: models/include/cloe/utility/frustum_culling.hpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+    if (distance >= frustum.clip_near && distance < frustum.clip_far) {
+      ret_val = true;
+    }
+  }
+  return ret_val;
+}
+
+}
+}
```
Comment (cassava) [inline models/include/cloe/utility/frustum_culling.hpp:156]: EOL

File: plugins/frustum_culling/src/frustum_culling_conf.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mats4) [reply plugins/frustum_culling/src/frustum_culling_conf.hpp]: didnt get the second comment.
changed to cloe::frustum_culling_plugin for now.

File: plugins/frustum_culling/.vscode/launch.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Debug frustum",
```
[the remark this answers is not in the document]
Comment (mats4) [reply plugins/frustum_culling/.vscode/launch.json]: i decided to add it (and renamed it as commented). in my opinion it is worth adding it, many developers use vscode and it helps them to start the debugging.

File: plugins/frustum_culling/.gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+build/
```
[the remark this answers is not in the document]
Comment (mats4) [reply plugins/frustum_culling/.gitignore]: deleted complete .gitignore file.

File: plugins/frustum_culling/readme.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+make export-all
+make -C meta package CONAN_OPTIONS="--build"
+```
+
+Then cd into plugins/frustum_culling
+and 
+``` make export package ```
+
+If you want a more verbose output on the test cases, then run: \
+``` make export package CONAN_OPTIONS="-o test_verbose=True" ```
```
[the remark this answers is not in the document]
Comment (mats4) [reply plugins/frustum_culling/readme.md]: the suggested alternative gave some problems.
i reworked the instructions, kindly double check.

File: models/include/cloe/utility/frustum_culling.hpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+#include <cmath> // for cos, sin, tan
+
+#include <fmt/format.h>                // for format
+#include <cloe/component/frustum.hpp>  // for Frustum
+#include <Eigen/Geometry>              // for Vector3d
+
+
+namespace cloe {
+namespace utility {
```
[the remark this answers is not in the document]
Comment (mats4) [reply models/include/cloe/utility/frustum_culling.hpp]: changed.

File: models/include/cloe/utility/frustum_culling.hpp (hunk 2 above)
Comment (mats4) [reply models/include/cloe/utility/frustum_culling.hpp]: changed.

File: models/src/cloe/utility/frustum_culling_test.cpp [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+ * \see  cloe/utility/frustum_culling.hpp
+ */
+
+#include <list>
+#include <tuple>
+
+#include <gtest/gtest.h>
+#include <cloe/utility/frustum_culling.hpp>
+
+TEST(rotate_point, rotate_point) {
```
Comment (cassava) [inline models/src/cloe/utility/frustum_culling_test.cpp:29]: Name tests after what is being modelled. So... `TEST(models_frustum_culling, ...)` in all places here. This makes it easier to find test source.

File: plugins/frustum_culling/src/frustum_culling_conf.hpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+struct MountPose : public fable::Confable {
+  Eigen::Isometry3d pose;
+  double x, y, z;
+  double roll, pitch, yaw;
+
+  void to_json(fable::Json& j) const override {
+    j = fable::Json{
+        {"pose", pose},
+    };
+  }
```
Comment (cassava) [inline plugins/frustum_culling/src/frustum_culling_conf.hpp:51]: Why only output pose?

**Review by cassava — APPROVED** [review]
Looks good! There are a few minor points that could still be addressed, and we need to see why CI is failing the ESMini build.

File: plugins/frustum_culling/src/frustum_culling_conf.hpp (hunk 9 above)
Comment (mats4) [reply plugins/frustum_culling/src/frustum_culling_conf.hpp:51]: i think the idea by Tobi was to internally work with Isometry3d class, while from the outside (yamls) being able to set the values x,y,z,roll,pitch,yaw.
So writing pose as well as x,y,z,roll,pitch,yaw to json would contain redundant information. 
Not sure if this is a good class design utilizing this "convert()" function and have redundant information in the class.
Im free to alternative ideas here, maybe there is a better mechanism than the current one? E.g. just store the pose somewhere else and not in the config_?
I think initially Tobi used "from_json" or something similar in order to do this the conversion but that did not work - it seemed to be never called and no conversion form xyz rpy to isometry3d happened.

File: models/src/cloe/utility/frustum_culling_test.cpp (hunk 8 above)
Comment (mats4) [reply models/src/cloe/utility/frustum_culling_test.cpp]: renamed.

File: plugins/frustum_culling/src/frustum_culling_conf.hpp (hunk 9 above)
Comment (cassava) [reply plugins/frustum_culling/src/frustum_culling_conf.hpp:51]: Ok, makes sense.

**Pull request merged**